Everything from Before Plus Tests

jrrose opened the pull request:
Because I suck at git, I closed the previous pull request and created a new one.

jrrose committed:
First commit - basic tools working

This patch adds all the initial commands planned for
the voicemail admin application. There are still some
major changes needed including to how system output
is done (need to move a good bit of stuff to use the
node-voicemail-logger system) as well as adding tests.

jrrose committed:
Second commit - add files that were forgotten

jrrose committed:
Add tests and fix all the things.

jrrose committed:
Only log errors to file, improve formatting.

Does a little cheeky hack to the logger so that console.log is used
to write error messages to standard IO while still preserving allowing
the logger to write to files.

samuelg commented on Gruntfile.js:
```diff
@@ -0,0 +1,65 @@
+        jshintrc: true
+      },
+      all: [
+        'Gruntfile.js',
```
missing app.js, test/mock_dal/*.js, test/mock_dal/helpers/*.js, and test/mock_dal/repositories/*.js

Adding these exposes 15 linting errors.

samuelg commented on app.js:
```diff
@@ -0,0 +1,48 @@
+/* Write error messages to the console as well as to the logger */
+logger.error = function () {
+  var args = Array.prototype.slice.call(arguments, 0);
+  console.log.apply(this, args);
```
should use console.error for errors.

samuelg commented:
For testing purposes I would intercept console.log. The output is interfering with reading the test results.

samuelg commented:
Test coverage shows the following:

Statements   : 57.93% ( 190/328 )
Branches     : 42.86% ( 60/140 )
Functions    : 61.84% ( 47/76 )
Lines        : 57.93% ( 190/328 )

I would aim for at least 75-80%, especially functions and statements. Unless you cover a lot of edge cases, branches may be lower than 75% which can be OK.

samuelg commented on README.md:
````diff
@@ -1,4 +1,11 @@
+
+```bash
+$ node app.js
+```
````
Some example usage might be good here. Even though the tool is good at guiding users, lack of documentation will drive people away.

samuelg commented on app.js:
```diff
@@ -0,0 +1,48 @@
+
+/* Increase level of streams that write to standard out for info/error
+ * loggers so that these streams are ignored. Bit of a hack. */
+for (var index in logger.streams) {
```
You can disable the console from the config.json file. No need to do this here.

samuelg commented on app.js:
```diff
@@ -0,0 +1,48 @@
+
+initializer.onStart(logger)
+.then(function (result) {
+  if (result === -1) {
```
ints as error values are not common in JavaScript. I would use the ability of promises to throw and catch proper error objects here.

samuelg commented on lib/voicemail-admin-init.js:
```diff
@@ -0,0 +1,76 @@
+
+  return fsReadFile('database.json')
+  .then(function (result) {
+    return 0;
```
throw an error object. The caller can use promise.catch to catch it.

samuelg commented on lib/voicemail-admin-init.js:
```diff
@@ -0,0 +1,76 @@
+  })
+  .catch(function (err) {
+    if (err.code !== 'ENOENT') {
+      return -1;
```
throw an error object. The caller can use promise.catch to catch it.

samuelg commented on lib/voicemail-admin-init.js:
```diff
@@ -0,0 +1,76 @@
+      console.log('%s saved', dbConfigName);
+      console.log('Run \'node createtables.js\' if the database is ' +
+                'uninitialized.');
+      return -1;
```
throw an error object. The caller can use promise.catch to catch it.

samuelg commented on lib/voicemail-admin-init.js:
```diff
@@ -0,0 +1,76 @@
+      }
+
+      console.log('%s saved', dbConfigName);
+      console.log('Run \'node createtables.js\' if the database is ' +
```
Couldn't we initialize the db for the user if it's not already?

samuelg commented on lib/voicemail-admin.js:
```diff
@@ -0,0 +1,771 @@
+var getPrompt = Q.denodeify(prompt.get);
+
+function invalidCommand(result) {
+  return Q.fcall(function () {
```
It looks like fcall is used a lot to turn non IO bound functions to promises. Is there a reason for this? If a function doesn't have any async IO in it, I'm not sure why you would need to wrap it in a promise.

jrrose replied:
I can't say I'm in love with the concept either, but I wrote the program in such a way that all commands have to return a promise. That way I can hold off on allowing the next prompt until after the promise is concluded and I don't have to deal with things like out of order operation completion and redrawing prompts and such when commands complete. I could probably clean it up a bit... I'll look into it.

samuelg replied:
There's something to be said for a consistent API. I was just curious that's all.

jrrose committed:
Address review feedback, raise test coverage.

Test coverage is now slightly above 80 for statements/lines
and above 90 for functions. Test coverage only covers
voicemail-admin.js, which is expected/intended since everything
else is just bootstrapping and prompting for setup.

samuelg commented on README.md:
````diff
@@ -1,4 +1,60 @@
+```bash
+$ git clone https://github.com/asterisk/node-voicemail-admin.git
+$ cd node-voicemail-admin
+$ npm install -g .
````
I would change this to npm install. npm install -g will install all the node-voicemail-admin dependencies globally which will pollute the global space and is not necessary to run the application from the checkout directory.

superscrat replied:
I agree with Sam. -g is good for things like installing grunt or yeoman, where you may want to be able to easily use it from your command prompt/shell/terminal. 

Of course, you can still install node packages locally, even if they are installed globally.

http://blog.nodejs.org/2011/03/23/npm-1-0-global-vs-local-installation
https://docs.npmjs.com/misc/faq

jrrose replied:
addressed

samuelg commented on app.js:
```diff
@@ -0,0 +1,52 @@
+  return app.create({logger: logger, dal: dal});
+})
+.catch(function (err) {
+  logger.error(err);
```
If this error means the application failed to start I would call process.exit(1); so the process is killed with the proper exit code.

_addressed_

samuelg commented on lib/voicemail-admin-init.js:
```diff
@@ -0,0 +1,77 @@
+
+    logger.info('%s saved', dbConfigName);
+    logger.info('Exit now and run \'node createtables.js\' if the database ' +
+                'is uninitialized.');
```
s/is uninitialized/has not been initialized/

_addressed_

samuelg commented on lib/voicemail-admin.js:
```diff
@@ -0,0 +1,779 @@
+ * Returns true, prints a cute little exit message
+ */
+function exit() {
+  logger.info('So long and thanks for all the fish'.cyan);
```
Do not remove that!

_left alone forever_

samuelg commented on lib/voicemail-admin.js:
```diff
@@ -0,0 +1,779 @@
+          for (var key in folders) {
+            var folder = folders[key];
+
+            if (key === 'add') {
```
Since you know what the folders collection object looks like, just check to see if the type of the key is a function instead, this is a bit hard to figure out without knowing what folders looks like. Also comment as to why this is necessary.

jrrose replied:
EDIT: Oh, wait, that's untrue.

EDIT2: Addressed

samuelg commented on lib/voicemail-admin.js:
```diff
@@ -0,0 +1,779 @@
+ *
+ * @param {object} associativeArray - associative array to cull junk from
+ */
+function removeUgly(associativeArray) {
```
May want to rename this method. It's not very descriptive.

jrrose replied:
Changed to 'removeNonValues' for now. Not sure if that's the best way of doing it, but eh... it seemed more correct than 'removeNonStrings'

samuelg commented on lib/voicemail-admin.js:
```diff
@@ -0,0 +1,779 @@
+        return invalidSyntax(entryComponents);
+      }
+
+      var mailboxKey = entryComponents[2].split('@');
```
This is another area where commander would help a lot and reduce the code necessary to validate inputs.

samuelg commented on lib/voicemail-admin.js:
```diff
@@ -0,0 +1,779 @@
+          }
+
+          var mailbox = dal.mailbox.create(mailboxNumber, holdContext);
+          mailbox.password = password;
```
You can pass these parameters as part of the call to create. Just pass an object as the third param with keys matching the param names.

_addressed_

samuelg commented:
Awesome test coverage! I also really like the overall setup with commands defined in a config file and command definition being mapped using prompt to JavaScript functions. The big thing here is really that a library like commander would simply your options parsing and also supports printing help and validating input.

Update: ignore Commander for now. We can file it as a future improvement. Having said that, it would change the way the admin operates from taking over the shell to requiring a base CLI command with sub commands from the shell for each command. Whether we want to make that change in the future or not is not clear to me.

jrrose committed:
Address current round of feedback sans Commander

jrrose committed:
forgot to add commands folder

samuelg commented on lib/commands/folder.js:
```diff
@@ -0,0 +1,217 @@
+ *
+ * @param {object} associativeArray - associative array to cull junk from
+ */
+function removeNonValues(associativeArray) {
```
Instead of defining this twice in the mailbox and folder commands, extract it into lib/commands/common.js or lib/commands/util.js.

samuelg commented on lib/commands/context.js:
```diff
@@ -0,0 +1,137 @@
+  logger = dependencies.logger;
+
+  return {
+    showContexts: function(chosenCommand, entryComponents) {
```
Just an idea to put in the future improvement bucket: the command implementations could have a comment above them that includes everything needed to generate the commands.json file using a grunt task. This way both the file needed for the CLI as well as the actual implementation would include the same documentation and adding/remove/editing commands could reflect changes in the commands file via a grunt task.

samuelg commented:
Other than extracting the common method in lib/commands/mailbox and lib/commands/folder I don't see anything to stop us from merging this. Once that method has been extracted I say we're good to merge.

jrrose committed:
Extract common command method

Previously removeNonValues was duplicated in lib/commands/mailbox.js and
lib/commands/folder.js - this patch extracts that function and puts it in
lib/commands/helpers/common.js

jrrose committed:
Oops, very minor change involving copyright date.

samuelg commented:
Looks good to me.

jrrose commented:
Merged as squished commit.
